integrations/webchat/examples/pre-post-chat-forms/client/react/src/FormContainer.js

```diff
@@ -0,0 +1,208 @@
+import { useCallback, useEffect, useRef, useState } from 'react';
+import ReactDOM from 'react-dom';
+import { PropTypes } from 'prop-types';
+import { PostChatForm } from './forms/PostChatForm';
+import { PreChatForm } from './forms/PreChatForm';
+import './FormContainer.css';
+
+/**
+ * This component is responsible for managing the pre-chat and post-chat forms. When mounted, it will add listeners
+ * for the appropriate window events so it can show the forms at the appropriate time.
+ */
+function FormContainer({ instance }) {
+  // This is the resolve function for resolving the promises returned by our open and close event handlers. Those
+  // promises will remain unresolved while the pre-chat or post-chat forms are being displayed and they will only be
+  // resolved once those forms are closed. Web chat will wait for each promise to be resolved before completing the
+  // open or close event.
+  const promiseResolveRef = useRef();
+
+  // The name that the user entered into the pre-chat form. We will save this so we can send it when the first message
+  // is sent. We also save this in session storage so it'll be available if the user reloads the page and starts a
+  // new Watson Assistant session.
+  const fullNameRef = useRef(sessionStorage.getItem('FULL_NAME'));
+
+  // This indicates if the user has sent a (non-welcome) message to the assistant. We use this so we only show the
+  // post-chat form if the user has actually interacted with the assistant.
+  const hasSentMessageRef = useRef(false);
```

Comment:
You use `useState` when you need a re-render to occur when the value changes. If you do not need a re-render to occur you should use a ref instead. A ref is simpler than the dual value+set-state you have to deal with with useState and it also won't triggered unnecessary renders. In this case the value is merely being stored so that it can be use by the call back functions. It is not actually used during render.

Reply:
Ok that makes sense - thank you for the explanation.